### PR TITLE
Add .fail() function for early judgement failure

### PR DESCRIPTION
**Pull request by dionyziz**

### CODE REVIEW

**Review by OrfeasLitos — COMMENTED** [review]
Also implement a `succeed()` function. Calling `this.fail(errStr)` in case of failure but returning `BaseJudge.PASS` in case of success is asymmetric.

**Comment by dionyziz** [comment]
I think the asymmetry here is fine. In case of failure, this should propagate up the call stack as an exception, but there's no corresponding control flow in case of success -- we require the judge to `return`, not `throw`.

**Comment by OrfeasLitos** [comment]
Is javascript OK with using errors for control flow? This isn't python, errors could be too expensive for us.

**Review by cnasikas — APPROVED** [review]

File: db/judge/BaseJudge.js [hunk 1]
```diff
@@ -59,11 +66,26 @@ class BaseJudge {
...
    * Returns false, if solution is not a valid solution
    * or integer, designating the grade of the solution
    */
-  judge (solution) {
+  async performJudgement (...params) {
+    try {
+      return await this.judge(...params)
+    }
+    catch (e) {
+      if (e instanceof IncorrectSolutionError) {
```
Comment (cnasikas) [inline db/judge/BaseJudge.js:82]: There is a problem with portability. Assignments do not throw an `IncorrectSolutionError` for the moment.

**Comment by OrfeasLitos** [comment]
Now that I read the code again, I'm not sure I agree with this usage of errors. It is not obvious that [this](https://github.com/cnasikas/blockchain-course-assignments/pull/21/files#diff-fc15a36468044cb97196ceb389eb864dR84) may throw. A reader can be misled to think that [`judge`](https://github.com/cnasikas/blockchain-course-assignments/pull/21/files#diff-fc15a36468044cb97196ceb389eb864dR59) always returns `BaseJudge.PASS`.

**Review by cnasikas — CHANGES_REQUESTED** [review]
Portability should be solved in my opinion.

**Comment by dionyziz** [comment]
@OrfeasLitos I don't understand your concern about errors being expensive. Why would exception be expensive in JS?

**Comment by dionyziz** [comment]
@OrfeasLitos Regarding the lack of clarity in the line of code you pointed to, this is the reason why the function was renamed to "*ensure*ContainsDesiredOutput". Do you think this is insufficient?

File: db/judge/BaseJudge.js (hunk 1 above)
Comment (dionyziz) [reply db/judge/BaseJudge.js:82]: This change is backwards compatible.

**Review by cnasikas — APPROVED** [review]

**Review by cnasikas — APPROVED** [review]

**Comment by OrfeasLitos** [comment]
> @OrfeasLitos I don't understand your concern about errors being expensive. Why would exception be expensive in JS?

The argument in favor of throwing is that once the judge decides that a solution is wrong, the execution path is very clear. It is not the case that throwing may lead to complex execution paths. This point also defeats my previous argument about speed, since each submission can cause at most one error.

The assymetry still bothers me a bit, but I'd be OK with it if we make clearer which parts of the code may throw. Regarding `ensureContainsDesiredOutput()`, we can add a `// may throw IncorrectSolutionError` comment over it. There may be a cleaner solution, but I can't come up with anything else that won't be too complex.

**Comment by dionyziz** [comment]
Sure, I can put that comment near `ensureContainsDesiredOutput()`, but this is in a different PR.

**Pull request merged**